diff -durN linux-2.6.0-test7-bk.orig/include/linux/a.out.h linux-2.6.0-test7-bk/include/linux/a.out.h
--- linux-2.6.0-test7-bk.orig/include/linux/a.out.h	2003-10-15 17:13:07.000000000 +0200
+++ linux-2.6.0-test7-bk/include/linux/a.out.h	2003-12-18 12:13:42.556335188 +0100
@@ -127,16 +127,7 @@
 #define SEGMENT_SIZE PAGE_SIZE
 #endif
 
-#ifdef linux
-#include <asm/page.h>
-#if defined(__i386__) || defined(__mc68000__)
 #define SEGMENT_SIZE	1024
-#else
-#ifndef SEGMENT_SIZE
-#define SEGMENT_SIZE	PAGE_SIZE
-#endif
-#endif
-#endif
 
 #define _N_SEGMENT_ROUND(x) (((x) + SEGMENT_SIZE - 1) & ~(SEGMENT_SIZE - 1))
 
