diff -ur linux-libc-headers-2.6.8.1.orig/include/asm-sparc/page.h linux-libc-headers-2.6.8.1/include/asm-sparc/page.h
--- linux-libc-headers-2.6.8.1.orig/include/asm-sparc/page.h	2004-08-29 21:57:38.000000000 +0200
+++ linux-libc-headers-2.6.8.1/include/asm-sparc/page.h	2004-09-20 20:32:04.000000000 +0200
@@ -8,7 +8,7 @@
 #ifndef _SPARC_PAGE_H
 #define _SPARC_PAGE_H
 
-#include <unistd.h>
+//#include <unistd.h>
 
 #define PAGE_SIZE ((unsigned long)getpagesize())
 #define PAGE_SHIFT ((PAGE_SIZE > 65536) ? -1 : ((unsigned long[]){12,13,14,-1,15,-1,-1,-1,16}[PAGE_SIZE>>13]))
