diff -urN linux-libc-headers-2.6.7.0.org/include/linux/pkt_cls.h linux-libc-headers-2.6.7.0/include/linux/pkt_cls.h
--- linux-libc-headers-2.6.7.0.org/include/linux/pkt_cls.h	2004-07-17 23:50:30.341860176 +0200
+++ linux-libc-headers-2.6.7.0/include/linux/pkt_cls.h	2004-07-17 23:51:41.025114688 +0200
@@ -61,6 +61,7 @@
 	__u32		val;
 	int		off;
 	int		offmask;
+	unsigned long   kcnt;
 };
 
 struct tc_u32_sel
@@ -76,6 +77,9 @@
 	short			hoff;
 	__u32			hmask;
 
+	unsigned long           rcnt;
+	unsigned long           rhit;
+
 	struct tc_u32_key	keys[0];
 };
 
