Index: linux-libc-headers/include/linux/msdos_fs.h
===================================================================
RCS file: /cvsroot/linux-libc-headers/include/linux/msdos_fs.h,v
retrieving revision 1.5
retrieving revision 1.6
diff -u -r1.5 -r1.6
--- linux-libc-headers/include/linux/msdos_fs.h	17 Jan 2004 22:43:07 -0000	1.5
+++ linux-libc-headers/include/linux/msdos_fs.h	19 Aug 2004 18:18:25 -0000	1.6
@@ -116,6 +116,8 @@
 #define CT_LE_W(v) cpu_to_le16(v)
 #define CT_LE_L(v) cpu_to_le32(v)
 
+#include <asm/types.h>
+
 struct fat_boot_sector {
 	__u8	ignored[3];	/* Boot strap short or near jump */
 	__u8	system_id[8];	/* Name - can be used to special case
