--- linux-libc-headers-2.6.5.0/include/linux/pkt_sched.h.orig	2004-04-19 23:13:51.000000000 +0200
+++ linux-libc-headers-2.6.5.0/include/linux/pkt_sched.h	2004-04-25 19:26:50.921597168 +0200
@@ -27,6 +27,8 @@
    Particular schedulers may have also their private records.
  */
 
+#include <linux/types.h>
+
 struct tc_stats
 {
 	__u64	bytes;			/* NUmber of enqueues bytes */
@@ -154,6 +156,13 @@
 
 /* SFQ section */
 
+enum
+{
+	TCA_SFQ_HASH_CLASSIC,
+	TCA_SFQ_HASH_DST,
+	TCA_SFQ_HASH_SRC,
+};
+
 struct tc_sfq_qopt
 {
 	unsigned	quantum;	/* Bytes per round allocated to flow */
