diff -uNr linux-libc-headers-2.6.11.0.orig/include/linux/console_splash.h linux-libc-headers-2.6.11.0/include/linux/console_splash.h
--- linux-libc-headers-2.6.11.0.orig/include/linux/console_splash.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-libc-headers-2.6.11.0/include/linux/console_splash.h	2005-03-13 23:23:02.000000000 +0100
@@ -0,0 +1,13 @@
+#ifndef _LINUX_CONSOLE_SPLASH_H_
+#define _LINUX_CONSOLE_SPLASH_H_ 1
+
+/* A structure used by the framebuffer splash code (drivers/video/fbsplash.c) */
+struct vc_splash {
+	__u8 bg_color;				/* The color that is to be treated as transparent */
+	__u8 state;				/* Current splash state: 0 = off, 1 = on */
+	__u16 tx, ty;				/* Top left corner coordinates of the text field */
+	__u16 twidth, theight;			/* Width and height of the text field */
+	char* theme;
+};
+
+#endif
diff -uNr linux-libc-headers-2.6.11.0.orig/include/linux/console_struct.h linux-libc-headers-2.6.11.0/include/linux/console_struct.h
--- linux-libc-headers-2.6.11.0.orig/include/linux/console_struct.h	2005-03-13 21:54:02.000000000 +0100
+++ linux-libc-headers-2.6.11.0/include/linux/console_struct.h	2005-03-13 23:23:02.000000000 +0100
@@ -12,6 +12,7 @@
 struct vt_struct;
 
 #define NPAR 16
+#include <linux/console_splash.h>
 
 struct vc_data {
 	unsigned short	vc_num;			/* Console number */
@@ -90,6 +91,8 @@
 	unsigned long	vc_uni_pagedir;
 	unsigned long	*vc_uni_pagedir_loc;  /* [!] Location of uni_pagedir variable for this console */
 	struct vt_struct *vc_vt;
+
+	struct vc_splash vc_splash;
 	/* additional information is in vt_kern.h */
 };
 
diff -uNr linux-libc-headers-2.6.11.0.orig/include/linux/fb.h linux-libc-headers-2.6.11.0/include/linux/fb.h
--- linux-libc-headers-2.6.11.0.orig/include/linux/fb.h	2005-01-08 15:03:32.000000000 +0100
+++ linux-libc-headers-2.6.11.0/include/linux/fb.h	2005-03-13 23:33:46.000000000 +0100
@@ -8,6 +8,13 @@
 #define FB_MAJOR		29
 #define FB_MAX			32	/* sufficient for now */
 
+struct fb_splash_iowrapper
+{
+	unsigned short vc;		/* Virtual console */
+	unsigned char origin;		/* Point of origin of the request */
+	void *data;
+};
+
 /* ioctls
    0x46 is 'F'								*/
 #define FBIOGET_VSCREENINFO	0x4600
@@ -32,6 +39,15 @@
 #define FBIOPUT_MODEINFO        0x4617
 #define FBIOGET_DISPINFO        0x4618
 
+#define FBIOSPLASH_SETCFG	_IOWR('F', 0x19, struct fb_splash_iowrapper)
+#define FBIOSPLASH_GETCFG	_IOR('F', 0x1A, struct fb_splash_iowrapper)
+#define FBIOSPLASH_SETSTATE	_IOWR('F', 0x1B, struct fb_splash_iowrapper)
+#define FBIOSPLASH_GETSTATE	_IOR('F', 0x1C, struct fb_splash_iowrapper)
+#define FBIOSPLASH_SETPIC 	_IOWR('F', 0x1D, struct fb_splash_iowrapper)
+
+#define FB_SPLASH_THEME_LEN		128	/* Maximum lenght of a theme name */
+#define FB_SPLASH_IO_ORIG_KERNEL	0	/* Kernel ioctl origin */
+#define FB_SPLASH_IO_ORIG_USER		1 	/* User ioctl origin */
 
 #define FB_TYPE_PACKED_PIXELS		0	/* Packed Pixels	*/
 #define FB_TYPE_PLANES			1	/* Non interleaved planes */
